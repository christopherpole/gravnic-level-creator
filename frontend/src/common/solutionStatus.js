import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

import TickIcon from 'common/icons/tickIcon';
import CrossIcon from 'common/icons/crossIcon';
import WarningIcon from 'common/icons/warningIcon';
import LoadingIcon from 'common/icons/loadingIcon';

export const Wrapper = styled.div`
  display: inline-flex;
  justify-content: center;
  align-items: center;
`;

export const SolutionStatusText = styled.p`
  white-space: nowrap;
  margin-left: calc(${props => props.theme.structureSpacing} / 2);
`;

export const SolutionStatus = ({ loading, error, solution, maxMoves }) => {
  if (loading) {
    return (
      <Wrapper>
        <LoadingIcon size={15} />
        <SolutionStatusText>Solving</SolutionStatusText>
      </Wrapper>
    );
  }

  if (error) {
    return (
      <Wrapper>
        <CrossIcon size={15} />
        <SolutionStatusText>An error has occured</SolutionStatusText>
      </Wrapper>
    );
  }

  if (solution && solution.length === 0) {
    return (
      <Wrapper>
        <WarningIcon color="yellow" />
        <SolutionStatusText>Solved by default</SolutionStatusText>
      </Wrapper>
    );
  }

  if (solution) {
    return (
      <Wrapper>
        <TickIcon />
        <SolutionStatusText>
          Solvable in {solution.length} move{solution.length !== 1 && 's'}!
        </SolutionStatusText>
      </Wrapper>
    );
  }

  if (solution === false) {
    return (
      <Wrapper>
        <CrossIcon />
        <SolutionStatusText>
          Unsolvable in {maxMoves} move{solution.length !== 1 && 's'}!
        </SolutionStatusText>
      </Wrapper>
    );
  }

  return (
    <Wrapper>
      <WarningIcon size={15} color="yellow" />
      <SolutionStatusText>Not yet solved</SolutionStatusText>
    </Wrapper>
  );
};

SolutionStatus.defaultProps = {
  loading: false,
  solution: null,
  maxMoves: null,
  error: null,
};

SolutionStatus.propTypes = {
  loading: PropTypes.bool,
  solution: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.string), PropTypes.bool]),
  maxMoves: PropTypes.number,
  error: PropTypes.bool,
};

export default SolutionStatus;
