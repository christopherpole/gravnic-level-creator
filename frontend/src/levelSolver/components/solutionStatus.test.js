import React from 'react';
import { MOVE_UP, MOVE_DOWN } from 'gravnic-game';
import { configure, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';

import { SolutionStatus } from './solutionStatus';

configure({ adapter: new Adapter() });

describe('<SolutionStatus />', () => {
  let props;

  beforeEach(() => {
    props = {
      solution: null,
      loading: false,
      error: false,
    };
  });

  it('Renders without exploding', () => {
    const level = shallow(<SolutionStatus {...props} />);

    expect(level).toHaveLength(1);
  });

  it('Matches the current snapshot', () => {
    const level = shallow(<SolutionStatus {...props} />);

    expect(toJson(level)).toMatchSnapshot();
  });

  it('Matches the current snapshot when loading', () => {
    const level = shallow(<SolutionStatus {...props} loading />);

    expect(toJson(level)).toMatchSnapshot();
  });

  it('Matches the current snapshot when a solution is present', () => {
    const level = shallow(<SolutionStatus {...props} solution={[MOVE_UP, MOVE_DOWN]} />);

    expect(toJson(level)).toMatchSnapshot();
  });

  it('Matches the current snapshot when a solution of length 0 is present', () => {
    const level = shallow(<SolutionStatus {...props} solution={[]} />);

    expect(toJson(level)).toMatchSnapshot();
  });

  it('Matches the current snapshot when a level is unsolvable', () => {
    const level = shallow(<SolutionStatus {...props} solution={false} maxMoves={5} />);

    expect(toJson(level)).toMatchSnapshot();
  });

  it('Matches the current snapshot when loading', () => {
    const level = shallow(<SolutionStatus {...props} loading />);

    expect(toJson(level)).toMatchSnapshot();
  });

  it('Matches the current snapshot when there is an error', () => {
    const level = shallow(<SolutionStatus {...props} error />);

    expect(toJson(level)).toMatchSnapshot();
  });
});
