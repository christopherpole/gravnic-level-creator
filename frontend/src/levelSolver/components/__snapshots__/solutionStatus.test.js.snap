// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SolutionStatus /> Matches the current snapshot 1`] = `
<styled.div>
  <WarningIcon
    color="yellow"
    size={15}
  />
  <styled.p>
    Not yet solved
  </styled.p>
</styled.div>
`;

exports[`<SolutionStatus /> Matches the current snapshot when a level is unsolvable 1`] = `
<styled.div>
  <CrossIcon
    fillColor="red"
    size={15}
  />
  <styled.p>
    Unsolvable in 
    5
     move
    s
    !
  </styled.p>
</styled.div>
`;

exports[`<SolutionStatus /> Matches the current snapshot when a solution is present 1`] = `
<styled.div>
  <TickIcon
    fillColor="green"
    size={15}
  />
  <styled.p>
    Solvable in 
    2
     move
    s
    !
  </styled.p>
</styled.div>
`;

exports[`<SolutionStatus /> Matches the current snapshot when a solution of length 0 is present 1`] = `
<styled.div>
  <WarningIcon
    color="yellow"
    size={15}
  />
  <styled.p>
    Solved by default
  </styled.p>
</styled.div>
`;

exports[`<SolutionStatus /> Matches the current snapshot when loading 1`] = `
<styled.div>
  <LoadingIcon
    color="white"
    size={15}
    speed={1}
  />
  <styled.p>
    Solving
  </styled.p>
</styled.div>
`;

exports[`<SolutionStatus /> Matches the current snapshot when loading 2`] = `
<styled.div>
  <LoadingIcon
    color="white"
    size={15}
    speed={1}
  />
  <styled.p>
    Solving
  </styled.p>
</styled.div>
`;

exports[`<SolutionStatus /> Matches the current snapshot when there is an error 1`] = `
<styled.div>
  <CrossIcon
    fillColor="red"
    size={15}
  />
  <styled.p>
    An error has occured
  </styled.p>
</styled.div>
`;
